File: 06-binary-search-trees/20250801-intro/ts/solution.test.ts
import { 
  Node, 
  createNode, 
  validateBST, 
  findMin, 
  findMax, 
  insert, 
  deleteNode, 
  search, 
  inorderTraversal, 
  preorderTraversal, 
  postorderTraversal, 
  height, 
  size 
} from './solution';

describe('BST Introduction Tests', () => {
  // Test 1: Node Creation
  test('createNode creates a node with correct value', () => {
    const node = createNode(10);
    expect(node).toBeDefined();
    expect(node.value).toBe(10);
    expect(node.left).toBeNull();
    expect(node.right).toBeNull();
  });

  // Test 2: BST Validation - Single Node
  test('validateBST returns true for single node', () => {
    const root = createNode(5);
    expect(validateBST(root)).toBe(true);
  });

  // Test 3: BST Validation - Valid Small BST
  test('validateBST returns true for valid small BST', () => {
    // Create a valid BST: 3 <- 5 -> 7
    const root = createNode(5);
    root.left = createNode(3);
    root.right = createNode(7);
    
    expect(validateBST(root)).toBe(true);
  });

  // Test 4: BST Validation - Invalid BST
  test('validateBST returns false for invalid BST', () => {
    // Create an invalid BST: 8 <- 5 -> 7 (left child > parent)
    const root = createNode(5);
    root.left = createNode(8);
    root.right = createNode(7);
    
    expect(validateBST(root)).toBe(false);
  });

  // Test 5: BST Validation - Empty Tree
  test('validateBST returns true for empty tree', () => {
    expect(validateBST(null)).toBe(true);
  });

  // Test 6: Find Min - Single Node
  test('findMin returns correct value for single node', () => {
    const root = createNode(10);
    expect(findMin(root)).toBe(10);
  });

  // Test 7: Find Min - Empty Tree
  test('findMin returns -1 for empty tree', () => {
    expect(findMin(null)).toBe(-1);
  });

  // Test 8: Find Max - Single Node
  test('findMax returns correct value for single node', () => {
    const root = createNode(15);
    expect(findMax(root)).toBe(15);
  });

  // Test 9: Find Max - Empty Tree
  test('findMax returns -1 for empty tree', () => {
    expect(findMax(null)).toBe(-1);
  });

  // Test 10: Find Min/Max - Complex Tree
  test('findMin and findMax work for complex tree', () => {
    // Create tree: 1 <- 3 <- 5 -> 7 -> 9
    const root = createNode(5);
    root.left = createNode(3);
    root.left.left = createNode(1);
    root.right = createNode(7);
    root.right.right = createNode(9);
    
    expect(findMin(root)).toBe(1);
    expect(findMax(root)).toBe(9);
  });

  // Additional Tests for Extended BST Features

  // Test 11: Insert Operation
  test('insert operation works correctly', () => {
    let root: Node | null = null;
    root = insert(root, 5);
    root = insert(root, 3);
    root = insert(root, 7);
    
    expect(root).not.toBeNull();
    expect(root!.value).toBe(5);
    expect(root!.left).not.toBeNull();
    expect(root!.left!.value).toBe(3);
    expect(root!.right).not.toBeNull();
    expect(root!.right!.value).toBe(7);
  });

  // Test 12: Search Operation
  test('search operation works correctly', () => {
    const root = createNode(5);
    root.left = createNode(3);
    root.right = createNode(7);
    
    expect(search(root, 5)).toBe(true);
    expect(search(root, 3)).toBe(true);
    expect(search(root, 7)).toBe(true);
    expect(search(root, 10)).toBe(false);
  });

  // Test 13: Delete Operation
  test('delete operation works correctly', () => {
    const root = createNode(5);
    root.left = createNode(3);
    root.right = createNode(7);
    
    const updatedRoot = deleteNode(root, 3);
    expect(search(updatedRoot, 3)).toBe(false);
    expect(search(updatedRoot, 5)).toBe(true);
    expect(search(updatedRoot, 7)).toBe(true);
  });

  // Test 14: Inorder Traversal
  test('inorder traversal returns sorted values', () => {
    const root = createNode(5);
    root.left = createNode(3);
    root.right = createNode(7);
    root.left.left = createNode(1);
    root.right.right = createNode(9);
    
    const result = inorderTraversal(root);
    expect(result).toEqual([1, 3, 5, 7, 9]);
  });

  // Test 15: Preorder Traversal
  test('preorder traversal returns root-first values', () => {
    const root = createNode(5);
    root.left = createNode(3);
    root.right = createNode(7);
    
    const result = preorderTraversal(root);
    expect(result).toEqual([5, 3, 7]);
  });

  // Test 16: Postorder Traversal
  test('postorder traversal returns children-first values', () => {
    const root = createNode(5);
    root.left = createNode(3);
    root.right = createNode(7);
    
    const result = postorderTraversal(root);
    expect(result).toEqual([3, 7, 5]);
  });

  // Test 17: Height Calculation
  test('height calculation works correctly', () => {
    const root = createNode(5);
    root.left = createNode(3);
    root.right = createNode(7);
    root.left.left = createNode(1);
    
    expect(height(root)).toBe(2);
    expect(height(null)).toBe(-1);
  });

  // Test 18: Size Calculation
  test('size calculation works correctly', () => {
    const root = createNode(5);
    root.left = createNode(3);
    root.right = createNode(7);
    root.left.left = createNode(1);
    
    expect(size(root)).toBe(4);
    expect(size(null)).toBe(0);
  });
});

File: 06-binary-search-trees/20250801-intro/go/solution_test.go
package main

import (
	"testing"
)

func TestCreateNode(t *testing.T) {
	node := CreateNode(10)
	if node == nil {
		t.Fatal("CreateNode should return a non-nil node")
	}
	// Additional checks will depend on your implementation
}

func TestValidateBST_SingleNode(t *testing.T) {
	root := CreateNode(5)
	if !ValidateBST(root) {
		t.Error("Single node should be a valid BST")
	}
}

func TestValidateBST_ValidSmallBST(t *testing.T) {
	// Create a small valid BST: 2 <- 5 -> 7
	root := CreateNode(5)
	// You'll need to manually set up the tree structure based on your implementation
	if !ValidateBST(root) {
		t.Error("Valid BST should return true")
	}
}

func TestValidateBST_InvalidBST(t *testing.T) {
	// Create an invalid BST where left child > parent
	// You'll need to set up an invalid tree structure
	// This test may need modification based on your implementation
}

func TestValidateBST_NilRoot(t *testing.T) {
	if !ValidateBST(nil) {
		t.Error("Empty tree (nil root) should be considered a valid BST")
	}
}

func TestFindMin_SingleNode(t *testing.T) {
	root := CreateNode(10)
	min := FindMin(root)
	if min != 10 {
		t.Errorf("Expected minimum to be 10, got %d", min)
	}
}

func TestFindMin_EmptyTree(t *testing.T) {
	min := FindMin(nil)
	if min != -1 {
		t.Errorf("Expected -1 for empty tree, got %d", min)
	}
}

func TestFindMax_SingleNode(t *testing.T) {
	root := CreateNode(15)
	max := FindMax(root)
	if max != 15 {
		t.Errorf("Expected maximum to be 15, got %d", max)
	}
}

func TestFindMax_EmptyTree(t *testing.T) {
	max := FindMax(nil)
	if max != -1 {
		t.Errorf("Expected -1 for empty tree, got %d", max)
	}
}

func TestFindMinMax_ComplexTree(t *testing.T) {
	// This test will need to be implemented based on your Node structure
	// Create a tree like: 1 <- 3 <- 5 -> 7 -> 9
	// Expected min: 1, max: 9
}